zetaclient/chains/bitcoin/observer/event.go

```diff
@@ -0,0 +1,233 @@
+	TxHash string
+}
+
+// IsProcessable checks if the inbound event is processable
+func (event *BTCInboundEvent) CheckProcessability() InboundProcessability {
+	// compliance check on sender and receiver addresses
+	if config.ContainRestrictedAddress(event.FromAddress, event.ToAddress) {
+		return InboundProcessabilityComplianceViolation
+	}
+
+	// compliance check on receiver, revert/abort addresses in standard memo
+	if event.MemoStd != nil {
+		if config.ContainRestrictedAddress(
+			event.MemoStd.Receiver.Hex(),
+			event.MemoStd.RevertOptions.RevertAddress,
+			event.MemoStd.RevertOptions.AbortAddress,
+		) {
+			return InboundProcessabilityComplianceViolation
+		}
+	}
+
+	// donation check
+	if bytes.Equal(event.MemoBytes, []byte(constant.DonationMessage)) {
+		return InboundProcessabilityDonation
+	}
+
+	return InboundProcessabilityGood
+}
+
+// DecodeEventMemoBytes decodes the memo bytes as either standard or legacy memo
+func (ob *Observer) DecodeEventMemoBytes(event *BTCInboundEvent) error {
+	var (
+		err      error
+		memoStd  *memo.InboundMemo
+		receiver ethcommon.Address
+	)
+
+	// skip decoding donation tx as it won't go through zetacore
+	if bytes.Equal(event.MemoBytes, []byte(constant.DonationMessage)) {
+		return nil
+	}
+
+	// try to decode the standard memo as the preferred format
+	// the standard memo is NOT enabled for Bitcoin mainnet
+	if ob.Chain().ChainId != chains.BitcoinMainnet.ChainId {
+		memoStd, err = memo.DecodeFromBytes(event.MemoBytes)
+	}
+
+	switch {
+	case memoStd != nil:
+		// skip memo that carries improper data
+		if err != nil {
+			return errors.Wrap(err, "standard memo contains improper data")
+		}
+
+		// NoAssetCall will be disabled for Bitcoin until full V2 support
+		if memoStd.OpCode == memo.OpCodeCall {
+			return errors.New("NoAssetCall is disabled")
+		}
+
+		// ensure the revert address is valid and supported
+		revertAddress := memoStd.RevertOptions.RevertAddress
+		if revertAddress != "" {
+			btcAddress, err := chains.DecodeBtcAddress(revertAddress, ob.Chain().ChainId)
+			if err != nil {
+				return errors.Wrapf(err, "invalid revert address in memo: %s", revertAddress)
+			}
+			if !chains.IsBtcAddressSupported(btcAddress) {
+				return fmt.Errorf("unsupported revert address in memo: %s", revertAddress)
+			}
+		}
+		event.MemoStd = memoStd
+		receiver = memoStd.Receiver
+	default:
+		// fallback to legacy memo if standard memo decoding fails
+		parsedAddress, _, err := memo.DecodeLegacyMemoHex(hex.EncodeToString(event.MemoBytes))
+		if err != nil { // never happens
+			return errors.Wrap(err, "invalid legacy memo")
+		}
+		receiver = parsedAddress
+	}
+
+	// ensure the receiver is valid
+	if crypto.IsEmptyAddress(receiver) {
+		return errors.New("got empty receiver address from memo")
+	}
+	event.ToAddress = receiver.Hex()
+
+	return nil
+}
+
+// CheckEventProcessability checks if the inbound event is processable
+func (ob *Observer) CheckEventProcessability(event *BTCInboundEvent) bool {
```

Comment:
```suggestion
func (ob *Observer) CheckEventProcessability(event BTCInboundEvent) bool {
```
Let's stop using pointers when we don't need those

Reply:
updated in https://github.com/zeta-chain/node/pull/3025/commits/7a2880bf21e47dfa10d0f3a9dd3756064237afbd